pkg/ratelimit/http.go

```diff
@@ -0,0 +1,69 @@
+// Copyright © 2021 The Things Network Foundation, The Things Industries B.V.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ratelimit
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"net/http"
+	"strings"
+)
+
+// HTTPVaryBy returns a rate limiting key from a *http.Request.
+type HTTPVaryBy func(*http.Request) string
+
+// HTTPRemoteIP returns a key that varies by remote IP, honoring X-Forwarded-For headers.
+func HTTPRemoteIP(r *http.Request) string {
+	if xff := r.Header.Get("x-forwarded-for"); xff != "" {
+		// TODO: only for trusted proxies
+		s := strings.Split(xff, ",")[0]
+		if s != "" {
+			return strings.Trim(s, " ")
+		}
+	}
```

Comment:
We already have the `ProxyHeaders` middleware that sets `X-Real-IP` when a trusted proxy sets those headers (and otherwise unsets it)